sagan/models/attention.py

```diff
@@ -1,39 +1,68 @@
 import tensorflow as tf
-from tensorflow import keras
-from tensorflow.keras import layers, initializers
+from tensorflow.keras import layers
+
+from sagan.models.spectral_norm import SpectralNorm
 
 
 class SelfAttention(layers.Layer):
-    def __init__(
-        self, input_dims, trainable=True, name=None, dtype=None, dynamic=False, **kwargs
-    ):
-        super().__init__(
-            trainable=trainable, name=name, dtype=dtype, dynamic=dynamic, **kwargs
-        )
-        self.query_convolution = layers.Conv2D(filters=input_dims // 8, kernel_size=1)
-        self.key_convolution = layers.Conv2D(filters=input_dims // 8, kernel_size=1)
-        self.value_convolution = layers.Conv2D(filters=input_dims // 8, kernel_size=1)
+    def __init__(self, **kwargs):
+        super(SelfAttention, self).__init__(**kwargs)
 
     def build(self, input_shape):
-        self.gamma = self.add_weight(
-            self.name + "_gamma", shape=(), initializer=initializers.Zeros()
+        batch_size, height, width, num_channels = input_shape
+        self.n_feats = height * width
+        self.query_conv = layers.Conv2D(
+            filters=num_channels // 8,
+            kernel_size=1,
+            padding="same",
+            kernel_constraint=SpectralNorm(),
+            name="Conv_Theta",
         )
-
-    def call(self, inputs, *args, **kwargs):
-        query = self.query_convolution(inputs)
-        key = self.key_convolution(inputs)
-        value = self.value_convolution(inputs)
-        query_dim = tf.shape(query)
-        batch_size, height, width = query_dim[0], query_dim[1], query_dim[2]
-        proj_query = tf.reshape(query, (batch_size, height * width, -1))
-        proj_key = tf.transpose(
-            tf.reshape(key, (batch_size, height * width, -1)), (0, 2, 1)
+        self.key_conv = layers.Conv2D(
+            filters=num_channels // 8,
+            kernel_size=1,
+            padding="same",
+            kernel_constraint=SpectralNorm(),
+            name="Conv_Phi",
         )
-        proj_value = tf.transpose(
-            tf.reshape(value, (batch_size, height * width, -1)), (0, 2, 1)
+        self.conv_g = layers.Conv2D(
+            filters=num_channels // 2,
+            kernel_size=1,
+            padding="same",
+            kernel_constraint=SpectralNorm(),
+            name="Conv_G",
         )
-        energy = tf.matmul(proj_query, proj_key)
-        attention = tf.nn.softmax(energy)
-        out = tf.matmul(proj_value, tf.transpose(attention, (0, 2, 1)))
-        out = tf.reshape(tf.transpose(out, (0, 2, 1)), (batch_size, height, width, -1))
-        return tf.add(tf.multiply(out, self.gamma), inputs), attention
+        self.value_conv = layers.Conv2D(
+            filters=num_channels,
+            kernel_size=1,
+            padding="same",
+            kernel_constraint=SpectralNorm(),
+            name="Conv_AttnG",
+        )
+        self.sigma = self.add_weight(
```

Comment:
Can you please explain this step?

Reply:
Assigning more weight to the non-local evidence.

https://arxiv.org/pdf/1805.08318.pdf - Page 3